src/main/react/app/containers/App/dateInput.js

```diff
@@ -0,0 +1,104 @@
+import DatePicker from "react-datepicker/es";
+import React from "react";
+import { MONTHS, ORDERS, PERIODS, WEEKDAYS } from "./constants";
+import { nextOccurrenceByYear, nextOccurrenceByMonth, format, formatDate } from "../../utils/date";
+
+export default function DateInput(props) {
+  let { field, name } = props;
+  let self = this;
+  let { repeatPeriod, repeat } = self.state;
+
+  function updateCheckbox(subField) {
+    return function(e) {
+      let changedState = {};
+      changedState[field] = { ...self.state[field] };
+      changedState[field][subField] = e.target.checked;
+      self.setState(changedState);
+    };
+  }
+
+  function updateDate() {
+    return function(e) {
+      let changedState = {};
+      changedState[field] = { ...self.state[field] };
+      changedState[field].date = e;
+      self.setState(changedState);
+    };
+  }
+
+  function updateDynamicDateData(subField) {
+    return function(e) {
+      let changedState = {};
+      changedState[field] = { ...self.state[field] };
+      changedState[field]["dynamicDate"] = { ...self.state[field]["dynamicDate"] };
+      changedState[field]["dynamicDate"][subField] = e.target.value;
+      self.setState(changedState);
+    };
+  }
+
+  let dateObject = self.state[field];
+
+  return <div>
+    <hr/>
+    <div className={"form-row"}>
+      <div className={"form-item-label"}>{name}</div>
+      <input type='checkbox' checked={dateObject.dynamic}
+             onChange={updateCheckbox("dynamic")}/>
+      <label className={"form-item-label"} style={{ width: "auto" }}>Dynamic</label>
+    </div>
+    <hr/>
+
+    {!dateObject.dynamic &&
+    <div className={"form-row form-row-indent"}>
+      <label style={{ marginRight: 10 }}>Date</label>
+      <DatePicker className={"input input-text"} selected={dateObject.date} placeholder="select"
+                  onChange={updateDate()}/>
+    </div>
+    }
+    {dateObject.dynamic && <div>
+      <div className={"form-row form-row-indent"} style={{ display: "flex", flexDirection: "row", lineHeight: "40px" }}>
+
+        {(repeatPeriod >= PERIODS.Month || !repeat) &&
+        <div className={"custom-control custom-control-inline"} style={{ paddingLeft: 0, marginRight: 0 }}>
```

Comment:
We could consider making React classes for each control. I'm not 100% convinced it's worth the effort, but it would clean up this code

Reply:
Sounds like a great idea and worth a try!